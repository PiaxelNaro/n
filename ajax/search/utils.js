// @flow
const mapResultDataObject = (res: { data: Array<any>, more_url?: string }) => {
    let apps;
    if (res.data && res.data.length > 0) {
        apps = res.data.map(app => ({
            id: app.product_id,
            icon: app.icon,
            company: app.publisher_name,
            name: app.product_name,
            link: app.target_url,
            store: app.market,
            flagCode: app.country_code,
            showStoreIcon: true,
        }));
        return {
            results: apps,
            moreUrl: res.more_url || '',
        };
    }
    return {
        results: [],
    };
};

export default mapResultDataObject;
