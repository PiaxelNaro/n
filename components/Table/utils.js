// @flow

export type Align = 'left' | 'center' | 'right';

export type LooseColumn = {
    key?: string,
    keys?: Array<string>,
    label?: any,
    tooltip?: any,
    subColumns?: Array<LooseColumn>,
    size?: Array<string> | string,
    align?: Array<Align> | Align,
    isSortable?: boolean,
};

export type Column = {
    key: string,
    keys: Array<string>,
    label: any,
    tooltip: any,
    subColumns: Array<Column>,
    size: Array<string>,
    isSortable: boolean,
    align: Array<Align>,
    headerAlign: Align,
};

export const getColumnColSpan = (col: Column) =>
    Math.max(
        1,
        col.keys.length,
        col.subColumns.reduce((m, { keys }) => m + keys.length, 0)
    );

export const getColumnRowSpan = (col: Column) => (col.subColumns.length ? 1 : 2);

export const normalizeColumn = (column: LooseColumn): Column => {
    const subColumns = column.subColumns ? column.subColumns.map(normalizeColumn) : [];
    let keys = [];
    if (!subColumns.length) {
        keys = column.keys || [];
        if (!keys.length) {
            keys = column.key ? [column.key] : [];
        }
    }
    const size = keys.map((_, i) => {
        if (Array.isArray(column.size)) {
            return i in column.size ? column.size[i] : 'medium';
        }

        return column.size ? column.size : 'medium';
    });

    const align = keys.map((_, i) => {
        if (Array.isArray(column.align)) {
            return i in column.align ? column.align[i] : 'left';
        }

        return column.align ? column.align : 'left';
    });

    let headerAlign = 'left';
    if (subColumns.length <= 1 && align.length === 1) {
        // follow column align if not composed column
        headerAlign = align[0];
    }

    let key = column.key || keys.join('|');
    if (subColumns.length) {
        // eslint-disable-next-line no-shadow
        key = subColumns.map(({ key }) => key).join('|');
    }

    return {
        key,
        isSortable: Boolean(column.isSortable),
        label: column.label,
        tooltip: column.tooltip,
        subColumns,
        keys,
        size,
        headerAlign,
        align,
    };
};

function getFirstHTMLChildren(tr: HTMLElement): void | HTMLElement {
    return Array.from(tr.children).find(node => Boolean(node.tagName));
}

export const createSizeMatrix = (tableScroller: ?HTMLElement) => {
    const table = tableScroller ? tableScroller.querySelector('table') : null;

    if (!tableScroller || !table) {
        return {
            width: 'auto',
            scrollerWidth: 'auto',
            rows: [],
        };
    }

    const mainTrs = Array.from(table.querySelectorAll('tr'));
    return {
        width: `${table.offsetWidth}px`,
        scrollerWidth: `${tableScroller.getBoundingClientRect().width}px`,
        rows: mainTrs.map((tr: HTMLElement) => {
            let cells = [];
            if (tr.parentNode && tr.parentNode.tagName === 'THEAD') {
                cells = Array.from(tr.children).map(cell => ({
                    width: `${cell.offsetWidth}px`,
                }));
            }

            // Chrome mess up rowspan heights, so we need to rely on the first child size if the
            // row has no height or no first child.
            // Here we won't get height from element's offsetHeight
            // because table cell size in Firefox will contain decimal number
            let rowHeight = tr.getBoundingClientRect().height;
            const firstChild = getFirstHTMLChildren(tr);
            if (
                rowHeight === 0 &&
                firstChild &&
                (firstChild.tagName === 'TH' || firstChild.tagName === 'TD')
            ) {
                rowHeight = firstChild.getBoundingClientRect().height;
            } else if (rowHeight !== 0 && !firstChild) {
                rowHeight = 0;
            }
            return {
                height: `${rowHeight}px`,
                cells,
            };
        }),
    };
};
