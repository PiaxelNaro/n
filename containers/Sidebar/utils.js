// @flow
import { createSelector } from 'reselect';
import { type I18nType } from '../../modules/i18n';
import { type UserType } from '../../modules/user';

// If user is Non-ANA, empty sub navs
const getANASidebar = (user: UserType, i18n: I18nType) => {
    if (user.isANA) {
        return [
            {
                id: 'c',
                title: '',
                items: [
                    {
                        id: 'c1',
                        title: i18n.gettext('帮助'),
                        href: '/help/',
                        isVisible: true,
                    },
                    {
                        id: 'c2',
                        title: i18n.gettext('课程'),
                        href: '/course',
                        isVisible: true,
                    },
                ],
            },
        ];
    }
    return null;
};

// Store PP locked when
// 1. user does not buy any store contract
// 2. user only buy store PYA or store PYP
const isStorePPLocked = (user: UserType) =>
    !user.isINT || !user.intAuth.store || !user.intAuth.store.pp;

// Store PYA invisible when
// 1. user does not buy any store contract
// 2. user is store PP only or store PYP only
const isStorePYAUser = (user: UserType) =>
    user.isINT && user.intAuth.store && user.intAuth.store.pya;

// Store PYP visible when
// user is store PYA or store PYP
const isStorePYPVisible = (user: UserType) =>
    user.isINT &&
    user.intAuth.store &&
    (user.intAuth.store.pyp || user.intAuth.store.pya);

// Store PYP locked when
// 1. user does not buy store contract
// 2. user is not store PYP
const isStorePYPLocked = (user: UserType) =>
    !user.isINT || !user.intAuth.store || !user.intAuth.store.pyp;

// Usage PP unlocked when user buy usage pp
const isUsagePPLocked = (user: UserType) =>
    !user.isINT || !user.intAuth.usage || !user.intAuth.usage.pp;

// Usage PYA show and unlocked if user buy usage pya
const isUsagePYAUser = (user: UserType) =>
    user.isINT && user.intAuth.usage && user.intAuth.usage.pya;

// Usage retention unlocked if user buy usage pp
const isUsagePPUser = (user: UserType) =>
    user.isINT && user.intAuth.usage && user.intAuth.usage.pp;

const isMarketAsoUser = (user: UserType) =>
    user.isINT && user.intAuth.marketing && user.intAuth.marketing.aso;

const isMarketAdvertisingUser = (user: UserType) =>
    user.isINT && user.intAuth.marketing && user.intAuth.marketing.advertising;

const getINTSidebar = (user: UserType, i18n: I18nType) =>
    [
        {
            id: 'a',
            title: i18n.gettext('排课'),
            items: [
                {
                    id: 'a1',
                    title: i18n.gettext('选课'),
                    href: '/#',
                    isVisible: true,
                    isLocked: !user.isINT,
                },
            ],
        },
        {
            id: 'b',
            title: i18n.gettext('查询'),
            items: [
                {
                    id: 'b1',
                    title: i18n.gettext('课程表'),
                    href: '/apps/ios/top-chart/',
                    isVisible: true,
                },
            ],
        }
    ].map(child => ({
        ...child,
        items: child.items.filter(item => item.isVisible),
    }));

const getSidebar = createSelector(
    user => user,
    (user, i18n) => i18n,
    (user: UserType, i18n: I18nType) => {
        const ANA_NAV = getANASidebar(user, i18n);
        const INT_NAV = getINTSidebar(user, i18n);
        return {
            current: 'a1',
            home: {
                id: 'home',
                title: i18n.gettext('首页'),
                href: '/home',
                isVisible: true,
            },
            platforms: [
                {
                    id: 'my_app_analytics',
                    title: i18n.gettext('学生功能'),
                    href: '/f',
                    icon: 'aa-connections',
                    items: ANA_NAV,
                },
                {
                    id: 'market_intelligence',
                    title: i18n.gettext('管理员功能'),
                    items: INT_NAV,
                },
            ],
        };
    }
);

export default getSidebar;
