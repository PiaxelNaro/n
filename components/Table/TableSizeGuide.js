// @flow
/* eslint react/no-array-index-key: 0 */

import React from 'react';
import styled from 'styled-components';
import { type Column } from './utils';

const TableGuide = styled.tbody`
    height: 0;

    .table-hide {
        border: none;
        padding: 0;
    }
`;

const SIZE_MAP = {
    auto: {
        weight: 0,
        minWidth: 'auto',
    },
    extraSmall: {
        weight: 0.6,
        minWidth: '60px',
    },
    small: {
        weight: 0.8,
        minWidth: '80px',
    },
    medium: {
        weight: 1.2,
        minWidth: '120px',
    },
    large: {
        weight: 2.6,
        minWidth: '260px',
    },
    extraLarge: {
        weight: 4,
        minWidth: '400px',
    },
};

const TableSizeGuide = ({ columns }: { columns: Array<Column> }) => {
    const sizes = columns.reduce((m, column) => {
        if (column.subColumns.length) {
            return m.concat(...column.subColumns.map(({ size }) => size));
        }
        return m.concat(column.size);
    }, []);
    const totalWeight = sizes.reduce((m, size) => m + SIZE_MAP[size].weight, 0);

    return (
        <TableGuide>
            <tr className="table-hide">
                {sizes.map((size, i) => {
                    const perc = SIZE_MAP[size].weight / totalWeight * 100;
                    return (
                        <td
                            className="table-hide"
                            key={i}
                            style={{
                                width: perc > 0 ? `${perc}%` : 'auto',
                                minWidth: SIZE_MAP[size].minWidth,
                            }}
                        />
                    );
                })}
            </tr>
        </TableGuide>
    );
};

export default TableSizeGuide;
