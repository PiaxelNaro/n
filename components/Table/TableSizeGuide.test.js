// @flow

import React from 'react';
import { shallow } from 'enzyme';
import TableSizeGuide from './TableSizeGuide';
import { normalizeColumn } from './utils';

describe('<TableSizeGuide />', () => {
    it('set weight and min-width to table cells', () => {
        const columns = [
            { key: '1', size: ['auto'] },
            { key: '2', size: ['extraSmall'] },
            { key: '3', size: ['small'] },
            { key: '4', size: ['medium'] },
            { key: '5', size: ['large'] },
            { key: '6', size: ['extraLarge'] },
            {
                subColumns: [
                    { key: '7', size: ['auto'] },
                    { key: '8', size: ['extraSmall'] },
                    { key: '9', size: ['small'] },
                    { key: '10', size: ['medium'] },
                    { key: '11', size: ['large'] },
                    { key: '12', size: ['extraLarge'] },
                ],
            },
        ].map(normalizeColumn);

        expect(shallow(<TableSizeGuide columns={columns} />)).toMatchSnapshot();
    });
});
