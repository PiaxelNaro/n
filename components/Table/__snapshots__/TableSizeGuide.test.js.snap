// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableSizeGuide /> set weight and min-width to table cells 1`] = `
<TableSizeGuide__TableGuide>
  <tr
    className="table-hide"
  >
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "auto",
          "width": "auto",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "60px",
          "width": "3.260869565217391%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "80px",
          "width": "4.347826086956522%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "120px",
          "width": "6.521739130434782%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "260px",
          "width": "14.130434782608697%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "400px",
          "width": "21.73913043478261%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "auto",
          "width": "auto",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "60px",
          "width": "3.260869565217391%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "80px",
          "width": "4.347826086956522%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "120px",
          "width": "6.521739130434782%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "260px",
          "width": "14.130434782608697%",
        }
      }
    />
    <td
      className="table-hide"
      style={
        Object {
          "minWidth": "400px",
          "width": "21.73913043478261%",
        }
      }
    />
  </tr>
</TableSizeGuide__TableGuide>
`;
